solidity/contracts/VRF.sol

```diff
@@ -0,0 +1,252 @@
+ *
+ * Returns true iff the proof can be verified as showing that _output was
+ * generated as mandated.
+ *
+ * See the invocation of verifyVRFProof in VRF.js, for an example.
+ *
+ * Design notes
+ * ------------
+ *
+ * An elliptic curve point is generally represented in the solidity code as a
+ * uint256[2], corresponding to its affine coordinates in GF(fieldSize).
+ *
+ * For the sake of efficiency, this implementation deviates from the spec in
+ * some minor ways:
+ *
+ * - Keccak hash rather than SHA256. This is because it's provided natively by
+ *   the EVM, and therefore costs much less gas. The impact on security should
+ *   be minor.
+ *
+ * - Alt-BN128 curve instead of P-256. Group operations are provided for this
+ *   curve as precompiled contracts, so they're much cheaper. Note, however,
+ *   that this a pairing-friendly curve, which may have weaker security. See the
+ *   conclusion to https://eprint.iacr.org/2016/1102.pdf , which estimates that
+ *   such a curve probably provides about 110 bits of security.
+ *
+ *   It should be possible to abuse ECRECOVER to use secp256k1 instead, which
+ *   would mitigate this weakness. See
+ *   https://github.com/1Address/ecsol/blob/master/contracts/EC.sol
+ *   https://ethresear.ch/t/you-can-kinda-abuse-ecrecover-to-do-ecmul-in-secp256k1-today/2384/6
+ *   But note that that's estimated to cost about 32,000 gas, vs about 40,000
+ *   gas as implemented here https://github.com/clearmatics/mobius/issues/69 ,
+ *   and hopefully the gas cost of this implementation will drop significantly
+ *   https://github.com/ethereum/EIPs/pull/1108
+ *
+ * - scalarFromCurve recursively hashes and takes the relevant hash bits until
+ *   it finds a point less than the group order. This results in uniform
+ *   sampling over the the possible values scalarFromCurve could take. The spec
+ *   recommends just uing the first hash output as a uint256, which is a biased
+ *   sample. See the zqHash function.
+ *
+ * - hashToCurve recursively hashes until it finds a curve x-ordinate. The spec
+ *   recommends that the initial input should be concatenated with a nonce and
+ *   then hashed, and this input should be rehashed with the nonce updated until
+ *   an x-ordinate is found. Recursive hashing is slightly more efficient. The
+ *   spec also recommends (https://tools.ietf.org/html/rfc8032#section-5.1.3 ,
+ *   by the specification of RS2ECP) that the x-ordinate should be rejected if
+ *   it is greater than the prime modulus. The modulus for Alt-BN128 is a couple
+ *   of bits shorter than 2**256-1, so hashToCurve masks those bits (see
+ *   zqHash.)
+ *
+ *   The spec also requires the y ordinate of the hashToCurve to be negated if y
+ *   is odd. See http://www.secg.org/sec1-v2.pdf#page=17 . This appears to be
+ *   purely for compositional purposes: They need to specify some way to
+ *   determine y given x, and happened to pick that one. Here, y is determined
+ *   as (x^3+3)^{(p+1)/4, instead.}
```

Comment:
Interesting. Is this an alternative to the additional v parameter in secp256k1 signatures?